container/apptabs.go

```diff
@@ -3,678 +3,312 @@ package container
-type TabItem struct {
-	Text    string
-	Icon    fyne.Resource
-	Content fyne.CanvasObject
+	baseTabs
+	// Deprecated: Use `OnSelected func(*TabItem)` instead.
+	OnChanged func(tab *TabItem)
 }
 
-// TabLocation is the location where the tabs of a tab container should be rendered
-//
-// Since: 1.4
-type TabLocation int
-
-// TabLocation values
-const (
-	TabLocationTop TabLocation = iota
-	TabLocationLeading
-	TabLocationBottom
-	TabLocationTrailing
-)
-
 // NewAppTabs creates a new tab container that allows the user to choose between different areas of an app.
 //
 // Since: 1.4
 func NewAppTabs(items ...*TabItem) *AppTabs {
-	tabs := &AppTabs{BaseWidget: widget.BaseWidget{}, Items: items, current: -1}
-	if len(items) > 0 {
-		// Current is first tab item
-		tabs.current = 0
+	tabs := &AppTabs{
+		baseTabs: baseTabs{
+			BaseWidget: widget.BaseWidget{},
+			current:    -1,
+		},
 	}
 	tabs.ExtendBaseWidget(tabs)
-
-	if tabs.mismatchedContent() {
-		internal.LogHint("AppTabs items should all have the same type of content (text, icons or both)")
-	}
-
+	tabs.SetItems(items)
 	return tabs
 }
 
-// NewTabItem creates a new item for a tabbed widget - each item specifies the content and a label for its tab.
-//
-// Since: 1.4
-func NewTabItem(text string, content fyne.CanvasObject) *TabItem {
-	return &TabItem{Text: text, Content: content}
-}
-
-// NewTabItemWithIcon creates a new item for a tabbed widget - each item specifies the content and a label with an icon for its tab.
-//
-// Since: 1.4
-func NewTabItemWithIcon(text string, icon fyne.Resource, content fyne.CanvasObject) *TabItem {
-	return &TabItem{Text: text, Icon: icon, Content: content}
-}
-
-// Append adds a new TabItem to the rightmost side of the tab panel
-func (c *AppTabs) Append(item *TabItem) {
-	c.SetItems(append(c.Items, item))
-}
-
 // CreateRenderer is a private method to Fyne which links this widget to its renderer
-func (c *AppTabs) CreateRenderer() fyne.WidgetRenderer {
-	c.ExtendBaseWidget(c)
-	r := &appTabsRenderer{line: canvas.NewRectangle(theme.ShadowColor()),
-		underline: canvas.NewRectangle(theme.PrimaryColor()), container: c}
-	r.updateTabs()
+func (t *AppTabs) CreateRenderer() fyne.WidgetRenderer {
+	t.ExtendBaseWidget(t)
+	r := &appTabsRenderer{
+		baseTabsRenderer: baseTabsRenderer{
+			bar:         &fyne.Container{},
+			divider:     canvas.NewRectangle(theme.ShadowColor()),
+			indicator:   canvas.NewRectangle(theme.PrimaryColor()),
+			buttonCache: make(map[*TabItem]*tabButton),
+		},
+		appTabs: t,
+	}
+	// Initially setup the tab bar to only show one tab, all others will be in overflow.
+	// When the widget is laid out, and we know the size, the tab bar will be updated to show as many as can fit.
+	r.updateTabs(1)
+	r.updateIndicator()
 	return r
 }
 
 // CurrentTab returns the currently selected TabItem.
-func (c *AppTabs) CurrentTab() *TabItem {
-	if c.current < 0 || c.current >= len(c.Items) {
+//
+// Deprecated: Use `AppTabs.Selection() *TabItem` instead.
```

Comment:
I thought that selection API change was to be a different PR?

Reply:
yes, removing the deprecations will be a separate PR

Reply:
What I meant is that moving to a new selection API should be a separate PR, as it is not core to getting this working.
Additionally the `Selection` was illustrated for multi-select towards the end of the discussion on this.
`Selected` seems more appropriate for single item selection, and matches the API elsewhere of `OnSelected`.
Our other widgets still use `Selected` until we agree on the larger item of work to move them all to a new API in preparation for multiselect.

Reply:
By "our other widgets" do you mean `widget.Select`? It does not implement the Selection API from (#1395) so I don't think it should not be used as a guiding principle. Actually the only widgets implementing #1395 are `widget.List`, `widget.Table`, and `widget.Tree` and neither #1395 nor any of the implementations have a `Selected` method.

I could deprecate `CurrentTab` and `CurrentTabIndex` and not provide an alternative, would that be better?

What selection model should DocTabs use? It would be bad to introduce an API in a new widget that will be deprecated. It would also be bad if the two tabbers had different selection APIs.

Reply:
Implemented `func Selected()`